presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+import com.google.common.collect.Sets;
+import io.prestosql.plugin.hive.metastore.Column;
+import io.prestosql.plugin.hive.metastore.Partition;
+import io.prestosql.plugin.hive.metastore.SemiTransactionalHiveMetastore;
+import io.prestosql.plugin.hive.metastore.Table;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.procedure.Procedure;
+import io.prestosql.spi.procedure.Procedure.Argument;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.invoke.MethodHandle;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import static io.prestosql.plugin.hive.HiveMetadata.PRESTO_QUERY_ID_NAME;
+import static io.prestosql.plugin.hive.HivePartitionManager.extractPartitionValues;
+import static io.prestosql.plugin.hive.HiveSessionProperties.isRespectTableFormat;
+import static io.prestosql.plugin.hive.PartitionStatistics.empty;
+import static io.prestosql.plugin.hive.metastore.StorageFormat.fromHiveStorageFormat;
+import static io.prestosql.spi.block.MethodHandleUtil.methodHandle;
+import static io.prestosql.spi.type.StandardTypes.BOOLEAN;
+import static io.prestosql.spi.type.StandardTypes.VARCHAR;
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class MsckProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle MSCK = methodHandle(
+            MsckProcedure.class,
+            "msck",
+            ConnectorSession.class,
+            String.class,
+            String.class,
+            Boolean.class,
+            Boolean.class);
+
+    private final Supplier<TransactionalMetadata> hiveMetadataFactory;
+    private final HdfsEnvironment hdfsEnvironment;
+
+    @Inject
+    public MsckProcedure(
+            Supplier<TransactionalMetadata> hiveMetadataFactory,
+            HdfsEnvironment hdfsEnvironment)
+    {
+        this.hiveMetadataFactory = requireNonNull(hiveMetadataFactory, "hiveMetadataFactory is null");
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "msck",
+                ImmutableList.of(
+                        new Argument("schema_name", VARCHAR),
+                        new Argument("table_name", VARCHAR),
+                        new Argument("add", BOOLEAN),
+                        new Argument("drop", BOOLEAN)),
+                MSCK.bindTo(this));
+    }
+
+    public void msck(ConnectorSession session, String schemaName, String tableName, Boolean add, Boolean drop)
```

Comment:
> Is it legal to call the procedure with `!add, !drop`?
> * it provides for a form of dry run, but since there is no return value, this is not overly useful

I have been thinking about it. Should we return the partition info?

Reply:
> * `@UsedByGeneratedCode`

Adding `@UsedByGeneratedCode` annotation would force `presto-hive` to depend on `presto-main`, and some other modules(e.g., `presto-parser`).

I am not sure if this is a good idea, or even if this is possible. I see the other Hive procedure(`CreateEmptyPartitionProcedure`) also does not add this annotation.

Reply:
> Adding `@UsedByGeneratedCode` annotation would force `presto-hive` to depend on `presto-main`, and some other modules(e.g., `presto-parser`).

thanks for checking